File: sort.py
"""
    SORT: A Simple, Online and Realtime Tracker
    Copyright (C) 2016-2020 Alex Bewley alex@bewley.ai

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
    
"""
#
from __future__ import print_function

import os
import numpy as np
import matplotlib
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt
import matplotlib.patches as patches
from skimage import io

import glob
import time
import argparse
from filterpy.kalman import KalmanFilter

np.random.seed(0)

#for hungarian algorithm : 헝가리안 알고리즘
def linear_assignment(cost_matrix):
  try:
    import lap
    _, x, y = lap.lapjv(cost_matrix, extend_cost=True)
    return np.array([[y[i],i] for i in x if i >= 0]) #
  except ImportError:
    from scipy.optimize import linear_sum_assignment
    x, y = linear_sum_assignment(cost_matrix)
    return np.array(list(zip(x, y)))


def iou_batch(bb_test, bb_gt):
  """
  From SORT: Computes IOU between two bboxes in the form [x1,y1,x2,y2]
  두 바운딩 박스 사이에 IOU 계산하기
  """
  bb_gt = np.expand_dims(bb_gt, 0)
  bb_test = np.expand_dims(bb_test, 1)
  
  xx1 = np.maximum(bb_test[..., 0], bb_gt[..., 0])
  yy1 = np.maximum(bb_test[..., 1], bb_gt[..., 1])
  xx2 = np.minimum(bb_test[..., 2], bb_gt[..., 2])
  yy2 = np.minimum(bb_test[..., 3], bb_gt[..., 3])
  w = np.maximum(0., xx2 - xx1)
  h = np.maximum(0., yy2 - yy1)
  wh = w * h
  o = wh / ((bb_test[..., 2] - bb_test[..., 0]) * (bb_test[..., 3] - bb_test[..., 1])                                      
    + (bb_gt[..., 2] - bb_gt[..., 0]) * (bb_gt[..., 3] - bb_gt[..., 1]) - wh)                                              
  return(o)  


def convert_bbox_to_z(bbox):
  """
  박스 정보를 아래와 같이 convert
  z = [center x,y 좌표 , 박스 크기, 너비/높이 비율]
  
  """
  w = bbox[2] - bbox[0] # x2 - x1
  h = bbox[3] - bbox[1] # y2 - y1
  x = bbox[0] + w/2. # 중심 좌표 x
  y = bbox[1] + h/2. # 중심 좌표 y
  s = w * h    # 바운딩 박스 스케일 (크기)
  r = w / float(h) # 가로 세로 비율
  return np.array([x, y, s, r]).reshape((4, 1))


def convert_x_to_bbox(x,score=None):
  """
  [x, y, s, r] -> [x1, y1, x2, y2]
  """
  w = np.sqrt(x[2] * x[3]) # s*r=w*h*w/h = w^2
  h = x[2] / w # w*h/w
  if(score==None): # score가 없다면
    return np.array([x[0]-w/2.,x[1]-h/2.,x[0]+w/2.,x[1]+h/2.]).reshape((1,4))
  else:
    return np.array([x[0]-w/2.,x[1]-h/2.,x[0]+w/2.,x[1]+h/2.,score]).reshape((1,5))


class KalmanBoxTracker(object):
  """
  This class represents the internal state of individual tracked objects observed as bbox.
  바운딩 박스로 관찰된 tracked objected의 상태를 나타냄
  tracking : 어떤 object의 물체의 위치를 찾는 과정
  tracker : 객체 정보
  """
  count = 0
  def __init__(self,bbox):
    """
    Initialises a tracker using initial bounding box.
    초기 바운딩 박스 값을 사용하여 tracker를 초기화 한다. (object 정보)
    """
    #define constant velocity model 
    #dim_x : size of the state vector : 상태 벡터의 수
    #dim_z : size of the measurement vector : 측정 벡터의 수
    # 시스템의 상태로부터 측정값을 추정하는 것이 목표


    self.kf = KalmanFilter(dim_x=7, dim_z=4) # 상태 벡터 7개 측정벡터 4개
    # x = [u, v, s, r, u', v', s']T => 개체의 무엇을 추적하고 있는지
    # 측정 벡터 [x1, y1, x2, y2] => 4개
    
    #  F : ndarray (dim_x, dim_x) /state transistion matrix
    #  F : 일부 식에서는 F가 아닌 A로 나와있기도 함. 시스템 행렬 - 상태 전이 행렬
    self.kf.F = np.array([[1,0,0,0,1,0,0],
                          [0,1,0,0,0,1,0],
                          [0,0,1,0,0,0,1],
                          [0,0,0,1,0,0,0],
                          [0,0,0,0,1,0,0],
                          [0,0,0,0,0,1,0],
                          [0,0,0,0,0,0,1]])
    # H : ndarray (dim_z, dim_x) /measurement function
    # H : 측정 기능 / 측정값과 상태변수의 관계
    self.kf.H = np.array([[1,0,0,0,0,0,0],
                          [0,1,0,0,0,0,0],
                          [0,0,1,0,0,0,0],
                          [0,0,0,1,0,0,0]])
    # R : ndarray (dim_z, dim_z), default eye(dim_x) /measurement uncertainty/noise
    # R : 측정 불확실성, 잡음 행렬
    self.kf.R[2:,2:] *= 10.
    # P : ndarray (dim_x, dim_x), default eye(dim_x) /covariance matrix
    # P : 공분산 행렬
    self.kf.P[4:,4:] *= 1000. #give high uncertainty to the unobservable initial velocities
    self.kf.P *= 10.
    # P가 크면 추정 오차가 크고 P가 작으면 추정 오차가 작다 => 높은 불확실성 부여 (초기)

    #Q : ndarray (dim_x, dim_x), default eye(dim_x) /Process uncertainty/noise
    #Q : 프로세스의 불확실성, 노이즈
    self.kf.Q[-1,-1] *= 0.01
    self.kf.Q[4:,4:] *= 0.01
    #x : ndarray (dim_x, 1), default = [0,0,0…0] /filter state estimate
    #상태변수 (관심있는 값)
    self.kf.x[:4] = convert_bbox_to_z(bbox) # [x,y,s,r]
    self.time_since_update = 0 # update부터 지난 시간
    self.id = KalmanBoxTracker.count # 오브젝트 아이디
    KalmanBoxTracker.count += 1 # count 값 1 증가 (object 개수?)
    self.history = [] # history 배열
    self.hits = 0 # update 수
    self.hit_streak = 0
    self.age = 0 # 예측 진행 수

  def update(self,bbox):
    """
    Updates the state vector with observed bbox.
    """
    self.time_since_update = 0
    self.history = []
    self.hits += 1 # update 수
    self.hit_streak += 1 # 이게 뭘까 ******************************
    self.kf.update(convert_bbox_to_z(bbox)) # update 함수 이용 
    # 칼만필터의 새 측정 z를 추가한다.

  def predict(self):
    """
    Advances the state vector and returns the predicted bounding box estimate.
    """
    if((self.kf.x[6]+self.kf.x[2])<=0): # 바운딩 박스의 크기 + 크기' <= 0
      self.kf.x[6] *= 0.0 # s'를 모두 0으로
    self.kf.predict() # predict 함수 이용
    self.age += 1 # 예측 진행
    if(self.time_since_update>0): # 방금 막 업데이트 되지 않음
      self.hit_streak = 0
    self.time_since_update += 1 # 업데이트 이후 지난 시간 1 증가
    self.history.append(convert_x_to_bbox(self.kf.x))
    # history 배열에 예측값 저장
    return self.history[-1] # 마지막 요소 반환

  def get_state(self):
    """
    Returns the current bounding box estimate.
    추정치 반환
    """
    return convert_x_to_bbox(self.kf.x)


def associate_detections_to_trackers(detections,trackers,iou_threshold = 0.3):
  """
  Assigns detections to tracked object (both represented as bounding boxes)

  Returns 3 lists of matches, unmatched_detections and unmatched_trackers
  """
  if(len(trackers)==0):
    return np.empty((0,2),dtype=int), np.arange(len(detections)), np.empty((0,5),dtype=int)

  iou_matrix = iou_batch(detections, trackers)

  if min(iou_matrix.shape) > 0:
    a = (iou_matrix > iou_threshold).astype(np.int32)
    if a.sum(1).max() == 1 and a.sum(0).max() == 1:
        matched_indices = np.stack(np.where(a), axis=1)
    else:
      matched_indices = linear_assignment(-iou_matrix)
  else:
    matched_indices = np.empty(shape=(0,2))

  unmatched_detections = []
  for d, det in enumerate(detections):
    if(d not in matched_indices[:,0]):
      unmatched_detections.append(d)
  unmatched_trackers = []
  for t, trk in enumerate(trackers):
    if(t not in matched_indices[:,1]):
      unmatched_trackers.append(t)

  #filter out matched with low IOU
  matches = []
  for m in matched_indices:
    if(iou_matrix[m[0], m[1]]<iou_threshold):
      unmatched_detections.append(m[0])
      unmatched_trackers.append(m[1])
    else:
      matches.append(m.reshape(1,2))
  if(len(matches)==0):
    matches = np.empty((0,2),dtype=int)
  else:
    matches = np.concatenate(matches,axis=0)

  return matches, np.array(unmatched_detections), np.array(unmatched_trackers)


class Sort(object):
  def __init__(self, max_age=1, min_hits=3, iou_threshold=0.3):
    """
    Sets key parameters for SORT
    """
    self.max_age = max_age
    self.min_hits = min_hits
    self.iou_threshold = iou_threshold
    self.trackers = []
    self.frame_count = 0

  def update(self, dets=np.empty((0, 5))):
    """
    Params:
      dets - a numpy array of detections in the format [[x1,y1,x2,y2,score],[x1,y1,x2,y2,score],...]
    Requires: this method must be called once for each frame even with empty detections (use np.empty((0, 5)) for frames without detections).
    Returns the a similar array, where the last column is the object ID.

    NOTE: The number of objects returned may differ from the number of detections provided.
    """
    self.frame_count += 1
    # get predicted locations from existing trackers.
    trks = np.zeros((len(self.trackers), 5))
    to_del = []
    ret = []
    for t, trk in enumerate(trks):
      pos = self.trackers[t].predict()[0]
      trk[:] = [pos[0], pos[1], pos[2], pos[3], 0]
      if np.any(np.isnan(pos)):
        to_del.append(t)
    trks = np.ma.compress_rows(np.ma.masked_invalid(trks))
    for t in reversed(to_del):
      self.trackers.pop(t)
    matched, unmatched_dets, unmatched_trks = associate_detections_to_trackers(dets,trks, self.iou_threshold)

    # update matched trackers with assigned detections
    for m in matched:
      self.trackers[m[1]].update(dets[m[0], :])

    # create and initialise new trackers for unmatched detections
    for i in unmatched_dets:
        trk = KalmanBoxTracker(dets[i,:])
        self.trackers.append(trk)
    i = len(self.trackers)
    for trk in reversed(self.trackers):
        d = trk.get_state()[0]
        if (trk.time_since_update < 1) and (trk.hit_streak >= self.min_hits or self.frame_count <= self.min_hits):
          ret.append(np.concatenate((d,[trk.id+1])).reshape(1,-1)) # +1 as MOT benchmark requires positive
        i -= 1
        # remove dead tracklet
        if(trk.time_since_update > self.max_age):
          self.trackers.pop(i)
    if(len(ret)>0):
      return np.concatenate(ret)
    return np.empty((0,5))

def parse_args():
    """Parse input arguments."""
    parser = argparse.ArgumentParser(description='SORT demo')
    parser.add_argument('--display', dest='display', help='Display online tracker output (slow) [False]',action='store_true')
    parser.add_argument("--seq_path", help="Path to detections.", type=str, default='data')
    parser.add_argument("--phase", help="Subdirectory in seq_path.", type=str, default='train')
    parser.add_argument("--max_age", 
                        help="Maximum number of frames to keep alive a track without associated detections.", 
                        type=int, default=1)
    parser.add_argument("--min_hits", 
                        help="Minimum number of associated detections before track is initialised.", 
                        type=int, default=3)
    parser.add_argument("--iou_threshold", help="Minimum IOU for match.", type=float, default=0.3)
    args = parser.parse_args()
    return args

if __name__ == '__main__':
  # all train
  args = parse_args()
  display = args.display
  phase = args.phase
  total_time = 0.0
  total_frames = 0
  colours = np.random.rand(32, 3) #used only for display
  if(display):
    if not os.path.exists('mot_benchmark'):
      print('\n\tERROR: mot_benchmark link not found!\n\n    Create a symbolic link to the MOT benchmark\n    (https://motchallenge.net/data/2D_MOT_2015/#download). E.g.:\n\n    $ ln -s /path/to/MOT2015_challenge/2DMOT2015 mot_benchmark\n\n')
      exit()
    plt.ion()
    fig = plt.figure()
    ax1 = fig.add_subplot(111, aspect='equal')

  if not os.path.exists('output'):
    os.makedirs('output')
  pattern = os.path.join(args.seq_path, phase, '*', 'det', 'det.txt')
  for seq_dets_fn in glob.glob(pattern):
    mot_tracker = Sort(max_age=args.max_age, 
                       min_hits=args.min_hits,
                       iou_threshold=args.iou_threshold) #create instance of the SORT tracker
    seq_dets = np.loadtxt(seq_dets_fn, delimiter=',')
    seq = seq_dets_fn[pattern.find('*'):].split(os.path.sep)[0]
    
    with open(os.path.join('output', '%s.txt'%(seq)),'w') as out_file:
      print("Processing %s."%(seq))
      for frame in range(int(seq_dets[:,0].max())):
        frame += 1 #detection and frame numbers begin at 1
        dets = seq_dets[seq_dets[:, 0]==frame, 2:7]
        dets[:, 2:4] += dets[:, 0:2] #convert to [x1,y1,w,h] to [x1,y1,x2,y2]
        total_frames += 1

        if(display):
          fn = os.path.join('mot_benchmark', phase, seq, 'img1', '%06d.jpg'%(frame))
          im =io.imread(fn)
          ax1.imshow(im)
          plt.title(seq + ' Tracked Targets')

        start_time = time.time()
        trackers = mot_tracker.update(dets)
        cycle_time = time.time() - start_time
        total_time += cycle_time

        for d in trackers:
          print('%d,%d,%.2f,%.2f,%.2f,%.2f,1,-1,-1,-1'%(frame,d[4],d[0],d[1],d[2]-d[0],d[3]-d[1]),file=out_file)
          if(display):
            d = d.astype(np.int32)
            ax1.add_patch(patches.Rectangle((d[0],d[1]),d[2]-d[0],d[3]-d[1],fill=False,lw=3,ec=colours[d[4]%32,:]))

        if(display):
          fig.canvas.flush_events()
          plt.draw()
          ax1.cla()

  print("Total Tracking took: %.3f seconds for %d frames or %.1f FPS" % (total_time, total_frames, total_frames / total_time))

  if(display):
    print("Note: to get real runtime results run without the option: --display")

File: sort-cpp/sort.cpp
///////////////////////////////////////////////////////////////////////////////
//  SORT: A Simple, Online and Realtime Tracker
//  
//  This is a C++ reimplementation of the open source tracker in
//  https://github.com/abewley/sort
//  Based on the work of Alex Bewley, alex@dynamicdetection.com, 2016
//
//  Cong Ma, mcximing@sina.cn, 2016
//  
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//  
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//  
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
///////////////////////////////////////////////////////////////////////////////


#include <iostream>
#include <fstream>
#include <iomanip> // to format image names using setw() and setfill()
#include <io.h>    // to check file existence using POSIX function access(). On Linux include <unistd.h>.
#include <set>

#include "Hungarian.h"
#include "KalmanTracker.h"

#include "opencv2/video/tracking.hpp"
#include "opencv2/highgui/highgui.hpp"

using namespace std;
using namespace cv;

typedef struct TrackingBox
{
	int frame;
	int id;
	Rect_<float> box;
}TrackingBox;


// Computes IOU between two bounding boxes
double GetIOU(Rect_<float> bb_test, Rect_<float> bb_gt)
{
	float in = (bb_test & bb_gt).area();
	float un = bb_test.area() + bb_gt.area() - in;

	if (un < DBL_EPSILON)
		return 0;

	return (double)(in / un);
}

float* equation(float y) {
    float x_range[2];
    x_range[0] = ((y - ROI[2][1]) * (ROI[0][0] - ROI[2][0]) / (ROI[0][1] - ROI[2][1])) + ROI[2][0];
    x_range[1] = ((y - ROI[1][1]) * (ROI[3][0] - ROI[1][0]) / (ROI[3][1] - ROI[1][1])) + ROI[1][0];

    return x_range;
}

bool isInROI(TrackingBox tb) {
   if (tb.box.y < 419)
      return false;

   float* x_range = equation(tb.box.y);

   if (tb.box.x < x_range[0] || tb.box.x > x_range[1]){
         return false;
   }
   return true;
}

void printbox(float x, float y, float w, float h)
{
    printf("x : %.3f\n", x);
    printf("y : %.3f\n", y);
    printf("w : %.3f\n", w);
    printf("h : %.3f\n", h);
}

// global variables for counting
#define CNUM 20
int total_frames = 0;
double total_time = 0.0;

float ROI[4][2] = { {812, 419}, {1088, 423}, {216, 1080}, {1698, 1080} };

float src_width = 1920;
float src_height = 1080;

void TestSORT(string seqName, bool display);

// 3. update across frames
int frame_count = 0;
int max_age = 1;
int min_hits = 3;
double iouThreshold = 0.3;
vector<KalmanTracker> trackers;
KalmanTracker::kf_count = 0; // tracking id relies on this, so we have to reset it in each seq.

// variables used in the for-loop
vector<Rect_<float>> predictedBoxes;
vector<vector<double>> iouMatrix;
vector<int> assignment;
set<int> unmatchedDetections;
set<int> unmatchedTrajectories;
set<int> allItems;
set<int> matchedItems;
vector<cv::Point> matchedPairs;
vector<TrackingBox> frameTrackingResult;
unsigned int trkNum = 0;
unsigned int detNum = 0;

double cycle_time = 0.0;
int64 start_time = 0;

//detdata 만드는 변수
string detLine;
istringstream ss;
vector<TrackingBox> getdetData;
float tpx, tpy, tpw, tph;

// detFrameData로 저장 변수
int maxFrame = 0;
vector<vector<TrackingBox>> getdetFrameData;
vector<TrackingBox> tempVec;

/*변수 꺼내오는 예제*/
void msgcallback4(const sort_VIP::Detector2DArray::ConstPtr& msg){

    total_frames++;
    frame_count++;

    unsigned int d_frame = msg->header.seq;
    printf("d_frame : %d\n", d_frame);
    printf("%d", sizeof(msg->detections));

    if (msg->detections.empty() == 0)
    {
        for (int i = 0; i < sizeof(msg->detections); i++)
        {
            TrackingBox tb;

            float cx = msg->detections[i].bbox.center.x;
            float cy = msg->detections[i].bbox.center.y;
            tpw = msg->detections[i].bbox.size_x;
            tph = msg->detections[i].bbox.size_y;
            tpx = cx - tpw / 2;
            tpy = cy - tph / 2;

            // d_id 받아와야함
            tb.frame.push_back(d_frame[i]);
            tb.id.push_back(d_id[i]);
            tb.box = Rect_<float>(Point_<float>(tpx, tpy), Point_<float>(tpx + tpw, tpy + tph));
            getdetData.push_back(tb);
        }

        // frame별 data담r기
        for (auto tb : getdetData)
            if (tb.frame == frame_count + 1) // frame num starts from 1
                tempVec.push_back(tb);
        getdetFrameData.push_back(tempVec);
        tempVec.clear();
    }
    TestSORT(getdetFrameData);
}

void TestSORT(vector<vector<TrackingBox>>& detFrameData)
{
	cout << "Processing " << seqName << "..." << endl;
	//////////////////////////////////////////////
	// main loop
	int fi = 0; fi < maxFrame; fi++;
    
	total_frames++;
    frame_count++;
    int frameIdx = fraframe_count-1;
    cout << frame_count << endl;

    // I used to count running time using clock(), but found it seems to conflict with cv::cvWaitkey(),
    // when they both exists, clock() can not get right result. Now I use cv::getTickCount() instead.
    start_time = getTickCount();

    if (trackers.size() == 0) // the first frame met
    {
        // initialize kalman trackers using first detections.
        // vector[0] == frame 1
        for (unsigned int i = 0; i < detFrameData[frameIdx].size(); i++)
        {
            KalmanTracker trk = KalmanTracker(detFrameData[frameIdx][i].box);
            trackers.push_back(trk);
        }
    }

    ///////////////////////////////////////
    // 3.1. get predicted locations from existing trackers.
    predictedBoxes.clear();

    for (auto it = trackers.begin(); it != trackers.end();)
    {
        Rect_<float> pBox = (*it).predict();
        if (pBox.x >= 0 && pBox.y >= 0)
        {
            predictedBoxes.push_back(pBox);
            it++;
        }
        else
        {
            it = trackers.erase(it);
        }
    }

    ///////////////////////////////////////
    // 3.2. associate detections to tracked object (both represented as bounding boxes)
    // dets : detFrameData[frame_count]
    trkNum = predictedBoxes.size();
    detNum = detFrameData[frameIdx].size();

    iouMatrix.clear();
    iouMatrix.resize(trkNum, vector<double>(detNum, 0));

    for (unsigned int i = 0; i < trkNum; i++) // compute iou matrix as a distance matrix
    {
        for (unsigned int j = 0; j < detNum; j++)
        {
            // use 1-iou because the hungarian algorithm computes a minimum-cost assignment.
            iouMatrix[i][j] = 1 - GetIOU(predictedBoxes[i], detFrameData[frame_count][j].box);
        }
    }

    // solve the assignment problem using hungarian algorithm.
    // the resulting assignment is [track(prediction) : detection], with len=preNum
    HungarianAlgorithm HungAlgo;
    assignment.clear();
    HungAlgo.Solve(iouMatrix, assignment);

    // find matches, unmatched_detections and unmatched_predictions
    unmatchedTrajectories.clear();
    unmatchedDetections.clear();
    allItems.clear();
    matchedItems.clear();

    if (detNum > trkNum) //	there are unmatched detections
    {
        for (unsigned int n = 0; n < detNum; n++)
            allItems.insert(n);

        for (unsigned int i = 0; i < trkNum; ++i)
            matchedItems.insert(assignment[i]);

        set_difference(allItems.begin(), allItems.end(),
                       matchedItems.begin(), matchedItems.end(),
                       insert_iterator<set<int>>(unmatchedDetections, unmatchedDetections.begin()));
    }
    else if (detNum < trkNum) // there are unmatched trajectory/predictions
    {
        for (unsigned int i = 0; i < trkNum; ++i)
            if (assignment[i] == -1) // unassigned label will be set as -1 in the assignment algorithm
                unmatchedTrajectories.insert(i);
    }
    else
        ;

    // filter out matched with low IOU
    matchedPairs.clear();
    for (unsigned int i = 0; i < trkNum; ++i)
    {
        if (assignment[i] == -1) // pass over invalid values
            continue;
        if (1 - iouMatrix[i][assignment[i]] < iouThreshold)
        {
            unmatchedTrajectories.insert(i);
            unmatchedDetections.insert(assignment[i]);
        }
        else
            matchedPairs.push_back(cv::Point(i, assignment[i]));
    }

    ///////////////////////////////////////
    // 3.3. updating trackers

    // update matched trackers with assigned detections.
    // each prediction is corresponding to a tracker
    int detIdx, trkIdx;
    for (unsigned int i = 0; i < matchedPairs.size(); i++)
    {
        trkIdx = matchedPairs[i].x;
        detIdx = matchedPairs[i].y;
        trackers[trkIdx].update(detFrameData[frameIdx][detIdx].box);
    }

    // create and initialise new trackers for unmatched detections
    for (auto umd : unmatchedDetections)
    {
        KalmanTracker tracker = KalmanTracker(detFrameData[frameIdx][umd].box);
        trackers.push_back(tracker);
    }

    // get trackers' output
    frameTrackingResult.clear();
    for (auto it = trackers.begin(); it != trackers.end();)
    {
        if (((*it).m_time_since_update < 1) &&
            ((*it).m_hit_streak >= min_hits || frame_count <= min_hits))
        {
            TrackingBox res;
            res.box = (*it).get_state();
            res.id = (*it).m_id + 1;
            res.frame = frame_count;
            frameTrackingResult.push_back(res);
            if (isInROI(res))
            {
                Perpective_trans_bbox.push_back(res);
            }
            it++;
        }
        else
            it++;

        // remove dead tracklet
        if (it != trackers.end() && (*it).m_time_since_update > max_age)
            it = trackers.erase(it);
    }

    // result
    cout << "///////////SORT RESULT" << endl;
    cout << "box info\n" << "x :" << Perpective_trans_bbox[frameIdx].box.x << endl;
    cout << "y : " << Perpective_trans_bbox[frameIdx].box.y << endl;
    cout << "width : " << Perpective_trans_bbox[frameIdx].box.width << endl;
    cout << "height : " << Perpective_trans_bbox[frameIdx].box.height << endl;            
}

int main(int argc, char **argv)
{
    /*rosros !! */
    ros::init(argc, argv, "sort_node");
    ros::NodeHandle n;
    ros::Subscriber sub4=n.subscribe<sort_VIP::Detector2DArray>("detections",100,msgcallback4);
    ros::spin();

    // -> perspective transformation
    sort_VIP::Detector2DArray::ConstPtr msg; //여기에 담으시면 됩니다 ^^ 
    ros::Publisher pub=n.advertise<sort_VIP::Detector2DArray>("sort_msg",100);
    //topic 이름이 sort_msg message 입니다 ^^ 
    sort_VIP::Detector2DArray::ConstPtr msg1;
    while(ros::ok){
        /*담으시면 됩니다 !!*/
        /*sort/msg 에 msg 만들고 그대로 빌드하면 message 생성되게 setting해놨습니다. 입맛 따라 */
        //msg1.sorted_cx=?
        //msg1.sorted_cy=?
        //msg1.sorted_width=?
        //msg1.sorted_height=?
        //msg1.삐리리=10
        //pub.publish(msg1);
    }

    return 0;
}
